ocaml/xapi/xapi_vdi.ml

```diff
@@ -141,18 +141,51 @@ let check_operation_error ~__context ?(sr_records=[]) ?(pbd_records=[]) ?(vbd_re
          | _ -> false)
       in
 
+      (* Disallow data_destroy when there is *any* VBD in the database at
+         all linking the VDI to a VM. *)
+      let operation_can_be_performed_with_detached_vbds = match op with
+        | `data_destroy -> false
+        | _ -> true
+      in
+
       (* NB RO vs RW sharing checks are done in xapi_vbd.ml *)
 
-      let blocked_by_attach =
-        if operation_can_be_performed_live
-        then false
-        else begin
-          if operation_can_be_performed_with_ro_attach
-          then (my_active_rw_vbd_records <> [])
-          else (my_active_vbd_records <> [])
-        end
+      let blocked_by_vbds =
+        let blocked_by_vbds =
+          if operation_can_be_performed_live
+          then false
+          else begin
+            if operation_can_be_performed_with_ro_attach
+            then (my_active_rw_vbd_records <> [])
+            else begin
+              if operation_can_be_performed_with_detached_vbds
+              then (my_active_vbd_records <> [])
+              else (my_vbd_records <> [])
+            end
+          end
+        in
+        let allow_vbds =
+          (* We use Safe_list.list to ignore exceptions due to invalid references that
+             could propagate to the message forwarding layer, which calls this
+             function to check for errors - these exceptions would prevent the
+             actual XenAPI function from being run. Checks called from the
+             message forwarding layer should not fail with an exception. *)
+          let true_for_all_active_vbds f = Safe_list.for_all f my_active_vbd_records in
+          match op with
+          | `list_changed_blocks ->
+            let vbd_connected_to_vm_snapshot vbd =
+              let vm = vbd.Db_actions.vBD_VM in
+              Db.is_valid_ref __context vm && (Db.VM.get_is_a_snapshot ~__context ~self:vm)
+            in
+            (* We allow list_changed_blocks on VDIs attached to snapshot VMs,
+               because VM.checkpoint may set the currently_attached fields of the
+               snapshot's VBDs to true, and this would block list_changed_blocks. *)
+            true_for_all_active_vbds vbd_connected_to_vm_snapshot
+          | _ -> false
```

Comment:
In the future we could do something more general like map each `op` to a filter function, and then say above:
```
if (filter my_..._vbd_records) <> []
```
for each branch of the if statement. But I'd say let's not overcomplicate it for now.